File: julia/p6.jl
function group_uniques(lines)
    vals = Set();
    for ln in lines
        for c in ln
            push!(vals, c);
        end
    end
    return length(vals);
end

function group_all_uniques(lines)
    vals = nothing;
    # new set for each line
    qset = nothing;

    for ln in lines
        qset = Set();
        for c in ln
            push!(qset, c);
        end
        if vals == nothing
            vals = qset;
        else
            vals = intersect(vals, qset);
        end
    end
    return length(vals);
end

function main()
    fname = "../data/p6.txt";
    lines = readlines(fname);
    tot = 0;
    tot_all = 0;
    curr_group = [];

    for ln in lines
        if length(ln) == 0
            tot += group_uniques(curr_group);
            tot_all += group_all_uniques(curr_group);
            curr_group = [];
        else
            push!(curr_group, ln);
        end
    end

    println("Part one: $tot");
    println("Part two: $tot_all");
end

main();

File: python/p6.py
def group_uniques(group):
    qs = set()

    for l in group:
        qs = qs.union(set(l))

    return len(qs)

def all_group_uniques(group):
    qs = None

    for l in group:
        if qs is None:
            qs = set(l)
        else:
            qs = qs.intersection(set(l))
    return len(qs)

def main():
    fname = '../data/p6.txt'

    with open(fname, 'r') as f:
        lines = f.read().splitlines()
        curr_group = []
        tot = 0
        tot_unique = 0

        for l in lines:
            if len(l) == 0:
                tot += group_uniques(curr_group)
                tot_unique += all_group_uniques(curr_group)
                curr_group = []
            else:
                curr_group.append(l)

        print("Part one: ", tot)
        print("Part one: ", tot_unique)

if __name__ == '__main__':
    main()
